### PR TITLE
lib/model: Change weak hash algo, add heuristic

### PR DESCRIPTION
**Pull request by AudriusButkevicius**

### CODE REVIEW

File: lib/model/rwfolder.go [hunk 1]
```diff
@@ -1216,7 +1220,7 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 		f.model.fmut.RUnlock()
 
 		var weakHashFinder *weakhash.Finder
-		if !f.DisableWeakHash {
+		if (len(state.file.Blocks)-state.have)*100/len(state.file.Blocks) >= f.WeakHashThresholdPct {
```
Comment (calmh) [inline lib/model/rwfolder.go]: this expression should definitely be a variable of some sort with a name

**Comment by calmh** [comment]
```
[RJBTF] 21:27:44 INFO: Single thread hash performance is 345 MB/s using minio/sha256-simd (345 MB/s using crypto/sha256).
...
[RJBTF] 21:27:58 VERBOSE: Scanning folder "default", 9% done (218.0 MiB/s)
```

still quite a lot of hashing performance we are losing here.

File: lib/model/rwfolder.go (hunk 1 above)
Comment (calmh) [inline lib/model/rwfolder.go]: Also, potential div by zero?

**Comment by AudriusButkevicius** [comment]
I guess we could go for xxhash, which would potentially raise that a bit, but then the rolling would be incredibly expensive

File: lib/model/rwfolder.go [hunk 2]
```diff
@@ -1216,7 +1220,8 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 		f.model.fmut.RUnlock()
 
 		var weakHashFinder *weakhash.Finder
-		if !f.DisableWeakHash {
+		blocksPercentChanged := (len(state.file.Blocks) - state.have) * 100 / len(state.file.Blocks)
```
Comment (calmh) [inline lib/model/rwfolder.go]: what about `len(state.file.Blockes) == 0`

Comment (AudriusButkevicius) [inline lib/model/rwfolder.go]: fixed

File: lib/model/rwfolder.go [hunk 3]
```diff
@@ -1216,7 +1220,12 @@ func (f *sendReceiveFolder) copierRoutine(in <-chan copyBlocksState, pullChan ch
 		f.model.fmut.RUnlock()
 
 		var weakHashFinder *weakhash.Finder
-		if !f.DisableWeakHash {
+		blocksPercentChanged := 0
+		if len(state.file.Blocks) != 0 {
```
Comment (calmh) [inline lib/model/rwfolder.go]:
```go
if tot := len(state.file.Blocks); tot > 0 {
    blocksPercentChanged = (tot - state.have) * 100 / tot
}
```

🚲 🏡

**Comment by calmh** [comment]
This is vast improvement on both side, it's good enough for now.

**Comment by AudriusButkevicius** [comment]
🚂 🚂

**Comment by calmh** [comment]
@st-review merge it yo

lib/model, lib/weakhash: Hash using adler32, add heuristic in puller

Adler32 is much faster, and the heuristic avoid the obvious cases where it will not help.

**Comment by st-review** [comment]
@calmh: Build status is `pending`. I'll wait until it goes green and then merge!

**Comment by calmh** [comment]
obtypo in the commit message, which will look like you did it 💯 😑

**Comment by st-review** [comment]
:ok_hand: Merged as 29d010ec0e7430089ed417581c31b2b503e57dd9. Thanks, @AudriusButkevicius!

**Pull request closed without merging**